packages/framework-provider-local/src/services/read-model-registry.ts

```diff
@@ -32,27 +37,67 @@ export class ReadModelRegistry {
     const queryPromise = new Promise((resolve, reject) =>
       cursor.exec((err, docs) => {
         if (err) reject(err)
-        else resolve(docs)
+        else {
+          docs.map((doc: ReadModelEnvelope) => {
+            doc.value.boosterMetadata!.optimisticConcurrencyValue = this.generateOptimisticConcurrencyValue(doc.value)
+          })
+          resolve(docs)
+        }
       })
     )
-
     return queryPromise as Promise<Array<ReadModelEnvelope>>
   }
 
-  public async store(readModel: ReadModelEnvelope): Promise<void> {
+  public async store(readModel: ReadModelEnvelope, expectedCurrentVersion: number): Promise<void> {
+    const readModelOptimistic: ReadModelEnvelope & { uniqueKey?: string } = readModel
+    readModelOptimistic.uniqueKey = `${readModel.typeName}_${readModel.value.id}_${readModel.value.boosterMetadata?.version}`
+    if (readModelOptimistic.value.boosterMetadata?.version === 1) {
+      return this.insert(readModel)
+    }
+    return this.update(readModelOptimistic, expectedCurrentVersion)
```

Comment:
So we have three values (I think) related to optimistic concurrency, but I don't fully get their purpose. We have:
- readModel.boosterMetadata.version (contains the read model version)
- readModel.uniqueKey (a unique key build from the version so that you can achieve the optimistic concurrency)
- readModel.boosterMetadata.optimisticConcurrencyValue (apparently, it also contains the read model version?)

Could you please clarify their purpose?

Reply:
You are right, the optimisticConcurrencyValue is not needed as it will be the version always. I will remove it and simplify it. The uniqueKey is, as you said, used to avoid **inserting** same ReadModel twice. The value of this field will be `<id>_<boosterMetadata.version>`. If the ReadModel doesn't exists in the database and there are two instances trying to insert the same ReadModel, then both of them will try to do it with version = 0. To be sure that only one of the insert finish successfully we will need to look up into the database before inserting to check if another instance inserted the same read model. As we don't are not able to lock rows or tables we need a unique key to ensure we will never insert a read model with the same and same version (in this case 0).